### ISSUE TITLE
Remove reliance on key order in "do"

### ISSUE THREAD

**Issue by matthias-pichler-warrify** [serverlessworkflow/specification#875]
**What would you like to be added**:

I propose to:
  1. only use the `continue` FlowDirective to continue a loop
  2. make `then` required in all tasks
  3. add a `start` field to the workflow that denotes the first task to execute

```yaml
document:
  dsl: '1.0.0-alpha1'
  namespace: test
  name: order-pet
  version: '1.0.0'
  title: Order Pet - 1.0.0
  summary: >
  # Order Pet - 1.0.0
  ## Table of Contents
  - [Description](#description)
  - [Requirements](#requirements)
  ## Description
  A sample workflow used to process an hypothetic pet order using the [PetStore API](https://petstore.swagger.io/)
  ## Requirements
  ### Secrets
   - my-oauth2-secret
use:
  authentications:
    petStoreOAuth2:
      oauth2: my-oauth2-secret
  functions:
    getAvailablePets:
      call: openapi
      with:
        document:
          uri: https://petstore.swagger.io/v2/swagger.json
        operation: findByStatus
        parameters:
          status: available
  secrets:
  - my-oauth2-secret
start: getAvailablePets
do:
  getAvailablePets:
    call: getAvailablePets
    output:
      from: "$input + { availablePets: [.[] | select(.category.name == "dog" and (.tags[] | .breed == $input.order.breed))] }"
    then: submitMatchesByMail
  submitMatchesByMail:
    call: http
    with:
      method: post
      endpoint:
        uri: https://fake.smtp.service.com/email/send
        authentication: petStoreOAuth2
      body:
        from: noreply@fake.petstore.com
        to: ${ .order.client.email }
        subject: Candidates for Adoption
        body: >
        Hello ${ .order.client.preferredDisplayName }!

        Following your interest to adopt a dog, here is a list of candidates that you might be interested in:

        ${ .pets | map("-\(.name)") | join("\n") }

        Please do not hesistate to contact us at info@fake.petstore.com if your have questions.

        Hope to hear from you soon!

        ----------------------------------------------------------------------------------------------
        DO NOT REPLY
        ----------------------------------------------------------------------------------------------
    then: end
```

**Why is this needed**:

Currently the task execution order is determined by the order in which the tasks appear in the "do" field. 
This is very convenient for writing _small_ workflows.
I however see two mayor drawbacks with this approach:

1. similar problems as with a "switch fallthrough". If one forgets a "then" field the execution might resume at a task that is different from the one I expected.

Another interesting case is that the `then` field is allowed top level in `switch` tasks. So The way I understand it, this would execute the `processElectronicOrder` task when the default case is reached because it sets a `continue` clause. 

```yaml
document:
  dsl: '1.0.0-alpha1'
  namespace: test
  name: sample-workflow
  version: '0.1.0'
do:
  processOrder:
    switch:
      case1:
        when: .orderType == "electronic"
        then: processElectronicOrder
      case2:
        when: .orderType == "physical"
        then: processPhysicalOrder
      default:
        then: continue
  processElectronicOrder:
    execute:
      sequentially:
        validatePayment: {...}
        fulfillOrder: {...}
    then: exit
  processPhysicalOrder:
    execute:
      sequentially:
        checkInventory: {...}
        packItems: {...}
        scheduleShipping: {...}
    then: exit
  handleUnknownOrderType:
    execute:
      sequentially:
        logWarning: {...}
        notifyAdmin: {...}
```

4. reliance on object keys forces implementers to alway process/store definitions as text blobs or strings. 
Some problems this causes:
  - One cannot store the definition as a JSON object in SQL or NoSQL databases as they usually do not preserve Object key order. Thus one cannot use built-in filtering in databases
  - When building an API to create workflows the definition has to be accepted as a string or file blob then parsed and validated with the JSON schema and then the parsed object cannot be used because key order is not preserved. This might look something like this in TypeScript:

```typescript
function parse(input: string): IDefinition {
  const obj: Workflow = JSON.parse(input);

  validate(obj);

  // parse document as map
  const map: Map<"do", Map<string, Task>> = yaml.parse(input, {
    mapAsMap: true,
  });
  // get task names in order of appearance
  const taskNames = Array.from(map.get("do")?.keys() ?? []);
  const tasks = new Map<string, Task>();
  taskNames.forEach((name) => {
    // get task as object instead of Map
    const task = obj.do[name];
    if (task) {
      tasks.set(name, task);
    }
  });

  return { ...obj, do: tasks, text: input };
}
```

and this only works because `Map` keys preserve order of insertion, assuming that `yaml` never changes the implementation that would change the insertion order.

3. when providing a UI to build/visualize workflows the logic to create a valid and correct workflow definition might become incredible complex because on has to be precise about the ordering of keys, handle exit and fallthrough cases etc...

**Comment by JBBianchi** [comment]
Note that `continue` might still be useful in loops (e.g.: `continue` triggers the next iteration, `exit` exists the loops, `end` ends the workflow from within the loop).

**Comment by matthias-pichler-warrify** [comment]
> Note that `continue` might still be useful in loops (e.g.: `continue` triggers the next iteration, `exit` exists the loops, `end` ends the workflow from within the loop).

Thanks, good to know 👍 I will updated the initial comment. On another note: could there be value in renaming `exit` to `break` since this more closely aligns to what programmers are already used to? Because I always confuse `end` and `exit` and thinking of exiting a process `process.exit(0)` makes `exit` even harder to remember correctly.

**Comment by JBBianchi** [comment]
> > Note that `continue` might still be useful in loops (e.g.: `continue` triggers the next iteration, `exit` exists the loops, `end` ends the workflow from within the loop).
> 
> Thanks, good to know 👍 I will updated the initial comment.
> 
> On another note: could there be value in renaming `exit` to `break` since this more closely aligns to what programmers are already used to? Because I always confuse `end` and `exit` and thinking of exiting a process `process.exit(0)` makes `exit` even harder to remember correctly.

I understand your point but I have 2 remarks:
- supposedly, the DSL should be accessible to non-developers (but, ofc, devs will probably be the main target)
- `break` kinda makes sense in the context of a loop, but does it really work for a branch ? We went for `exit` because it felt more natural to "exit" the branch than "breaking" it. It's still very opinionated and subject to discussions.

**Comment by matthias-pichler-warrify** [comment]
> > > Note that `continue` might still be useful in loops (e.g.: `continue` triggers the next iteration, `exit` exists the loops, `end` ends the workflow from within the loop).
> > 
> > 
> > Thanks, good to know 👍 I will updated the initial comment.
> > On another note: could there be value in renaming `exit` to `break` since this more closely aligns to what programmers are already used to? Because I always confuse `end` and `exit` and thinking of exiting a process `process.exit(0)` makes `exit` even harder to remember correctly.
> 
> I understand your point but I have 2 remarks:
> 
> * supposedly, the DSL should be accessible to non-developers (but, ofc, devs will probably be the main target)
> * `break` kinda makes sense in the context of a loop, but does it really work for a branch ? We went for `exit` because it felt more natural to "exit" the branch than "breaking" it. It's still very opinionated and subject to discussions.

I completely understand. I am also unsure about that. Exiting a branch/loop is also how I've come to remember it now.
Interestingly `exit` and `break` seem to align 1 to 1 (loops & switch statements). 

But yeah lot of overloaded terms in general.

**Comment by fjtirado** [comment]
I think one of the coolest features of the new DSL is the possibiiltiy to write sequential workflows with very few words. See also this somehow related [dicussion](https://github.com/serverlessworkflow/specification/discussions/872)
If we remove that reliance in order I feel we were losing a pretty nice feature. Of the three issues, two of them (sorting and UI)  are related with implementation difficulties that can be addressed. 
This leave us with the the switch, which maybe can be adressed by changing it, so once you start a switch, everything has to be nested inside.

**Comment by JBBianchi** [comment]
A possibility would be to add a parameter in the document object, like `executionSequence` with values like `strict` or `loose`. Loose would be the default, current behavior, and strict would require to provide a `start` and mandatory `then`.

**Comment by matthias-pichler-warrify** [comment]
> I think one of the coolest features of the new DSL is the possibiiltiy to write sequential workflows with very few words. See also this somehow related [dicussion](https://github.com/serverlessworkflow/specification/discussions/872) If we remove that reliance in order I feel we were losing a pretty nice feature.

I agree that for purely sequential workflows this is a very nice feature. However I personally do not think it is worth the footgun. I have two scenarios in mind here:

1. having worked with many workflow languages (Pipedream, Twilio Workflows, AWS Step Functions, Zapier, Cadence, Temporal and having implemented a custom workflow DSL before) the number of times a purely sequential workflows was enough is near zero. (A single API call would already require to handle a 404). Some might have started out this way but very few stayed that way,  which leads me to my seconds point:

2. evolving sequential flows:

Assume I have a sequential workflow that `doesStuff`:

```yaml
do:
  doStuff1: {}
  doStuff2: {}
```

when I run into issues where a even number is required I might add a check

```yaml
do:
  checkParity:
    isOdd:
      when: ${ .x % 2 == 1 }
      then: double
   default:
     then: continue
  double:
    set:
      x: ${ x * 2}
  doStuff1: {}
  doStuff2: {}
```

I would immediately introduce a bug because `x` is now always doubled. And I would scratch my head about where I should even place `double` for it to work.

I believe it would be easier to update a single reference to point to the newly added task than to go around and suddenly have to add then-references everywhere.

> Of the three issues, two of them (sorting and UI) are related with implementation difficulties that can be addressed.

Both the [JSON spec](https://datatracker.ietf.org/doc/html/rfc8259#section-1) as well as the [YAML spec](https://yaml.org/spec/1.2.2/#3221-mapping-key-order) define maps/objects as "unordered". So as long as YAML and/or JSON is chose as the definition language it seems wrong to me to go against these standards.
I believe this makes it incredibly hard for any implementor since basically no existing parsing library can be used. It would require everyone to implement a custom yaml/json parser. 

> This leave us with the the switch, which maybe can be adressed by changing it, so once you start a switch, everything has to be nested inside.

I think this would be even worse. Because then there would be no way for me to merge multiple branches of a switch task. Then I'd have to either copy and paste all tasks which makes it very verbose and unmaintainable or I'd have to create a completely separate workflow that all branches invoke.

**Comment by cdavernas** [comment]
> I would immediately introduce a bug because x is now always doubled. And I would scratch my head about where I should even place double for it to work.

No, this is not the intended behavior. The 'next' of the switch task is only performed if no default case has been supplied.

**Comment by matthias-pichler-warrify** [comment]
> > I would immediately introduce a bug because x is now always doubled. And I would scratch my head about where I should even place double for it to work.
> 
> No, this is not the intended behavior. The `next` of the switch task is only performed if no default case has been supplied. Your example would therefore work as expected.

but what about a default with `then: continue`?

**Comment by cdavernas** [comment]
> Having a required then field would allow me to add tasks anywhere in the execution graph by simply changing a single existing then property to point to the new subgraph.

That is true in theory, but in practice, I never used a serializer implementation which did not preserve the definition order. Plus, that would then be theoretically be true for lists, which should then be indexed thanks to some sort of additional property.

**Comment by matthias-pichler-warrify** [comment]
> > Having a required then field would allow me to add tasks anywhere in the execution graph by simply changing a single existing then property to point to the new subgraph.
> 
> That is true in theory, but in practice, I never used a serializer implementation which did not preserve the definition order. Plus, that would then be theoretically be true for lists, which should then be indexed thanks to some sort of additional property.

Arrays are defined with order in both specs and parsers respect that.

**Comment by cdavernas** [comment]
> Arrays are defined with order in both specs and parsers respect that.

I overlooked that. 

Refering to me previous comment, though, have you ever faced the case where a serializer did NOT preserve the order?

**Comment by matthias-pichler-warrify** [comment]
> > Arrays are defined with order in both specs and parsers respect that.
> 
> I overlooked that.
> 
> Refering to me previous comment, though, have you ever faced the case where a serializer did NOT preserve the order?

Yes: any parser in Javascript/Typescript that uses a Object representation for mappings (most do, some have support for a map)

also pyyaml uses `dict` by default and one has to subclass with an OrderedDict

**Comment by cdavernas** [comment]
Well, if there's no way around it, I believe you should made your point. Unhappily.

**Comment by cdavernas** [comment]
An alternative would be to do what Google Workflows do, thus solving everyone's problems:

```
do:
- task1:
    call: http
- task2:
    call: http
```

**Comment by matthias-pichler-warrify** [comment]
> An alternative would be to do what Google Workflows do, thus solving everyone's problems:
> 
> ```yaml
> do:
> - task1:
>     call: http
> - task2:
>     call: http
> ```
> 
> We ensure proper ordering while keeping brievity.

That would work for me 👍 My goal really is to not have to bend over backwards to pass the definition as a string everywhere having to parse it every time and be able to send it via our API as JSON and store it in out DB as a JSON Blob instead of string

**Comment by cdavernas** [comment]
> That would work for me 👍

Great!

> My goal really is to not have to bend over backwards

I know. And you saved us from a painful realisation in a near future. Thanks again!

Wanna take care of the PR to address it?

**Comment by fjtirado** [comment]
If we are going to use an array, I guess we might introduce names for task after all. isnt it?
See https://github.com/serverlessworkflow/specification/issues/849

**Comment by matthias-pichler-warrify** [comment]
> If we are going to use an array, I guess we might introduce names for task after all. isnt it? See #849

that makes sense 👍

**Comment by cdavernas** [comment]
> If we are going to use an array, I guess we might introduce names for task after all. isnt it?
See https://github.com/serverlessworkflow/specification/issues/849

@fjtirado I don't know how others feel about it, but I'd prefer keeping the name in the path, therefore using the solution I proposed, like google does (otherwise, would be the same than going back to where we were).

**Comment by fjtirado** [comment]
@cdavernas I think we need to compare the resulting schemas. 
If there is not name the object can be identified univocally by its position in the array.

**Comment by fjtirado** [comment]
Btw, when using jackson the map sorting is not an issue. It is using a linkedhashmap that preserve the order.

**Comment by fjtirado** [comment]
But to be honest I prefer an array with an optional property name. And that the do schema, as proposed [here]( https://github.com/serverlessworkflow/specification/discussions/872), is the same both for main flow, loops and retries

**Comment by cdavernas** [comment]
> @cdavernas I think we need to compare the resulting schemas.
If there is not name the object can be identified univocally by its position in the array.

Sure, whatever you guys want is fine. I'm just sad we have to go down that path again 😭 

> Btw, when using jackson the map sorting is not an issue. It is using a linkedhashmap that preserve the order. Im saying that because we can still keep the map if we want. The concept of keeping order in map is not an alien thing (might not the default for python, but it does exist in Java and is the default for one of the most popular jackson deserializers)

No, that's not acceptable to say that, because the libs we use work fine that way, we should discard the issue with others, specially when we are speaking of some of the most popular languages on earth, `js` and `python`. As much as I hate it, we are forced I believe to perform that change.

**Comment by fjtirado** [comment]
Once we renounce to the map thing because constraints on "popular" programming languages that are not really programming languages ;), I think we should take the the json schema that is easier to parse (in other words, the json schema with the minimun number of anyOf/oneOf constructs)
That why I fell that, to decide if we go with name as value of a property called name or name as propery itself, we should take a look to the resulting schemas.

**Comment by ricardozanini** [comment]
oh, it's hard to follow all the discussions at once. Nice to see some momentum and engagement, thou!

I vote for the proposal that I can do `do/name-of-task/...` in JSON pointers/path. I'd hate to be forced to use indexes again to map objects within the DSL.

**Comment by fjtirado** [comment]
@ricardozanini I need to see the schema that allows doing that when using an array

**Comment by matthias-pichler-warrify** [comment]
> Once we renounce to the map thing because constraints on "popular" programming languages that are not really programming languages ;), I think we should take the the json schema that is easier to parse (in other words, the json schema with the minimun number of anyOf/oneOf constructs) That why I feel that, to decide if we go with name as value of a property called name or name as propery itself, we should take a look to the resulting schemas.

The name as property vs not has one level of nesting difference I believe

```yaml
do:
  type: array
  items: 
    $ref: "$defs/Task"
$defs:
  Task:
    type: object
    properties: 
       name:
         type: string
       input: ...
       ...
    additionalProperties:
      oneOf:
        - $ref: $defs/CallTask
        - $ref: $defs/SwitchTask
        -  ...
```

or

```yaml
do:
  type: array
  items: 
    $ref: "$defs/Task"
$defs:
  Task:
    type: object
    maxProperties: 1
    minProperties: 1
    additionalProperties:
      type: object
      properties: 
         input: ...
         ...
      oneOf:
        - $ref: $defs/CallTask
        - $ref: $defs/SwitchTask
        -  ...
```

**Comment by matthias-pichler-warrify** [comment]
> oh, it's hard to follow all the discussions at once. Nice to see some momentum and engagement, thou!
> 
> I vote for the proposal that I can do `do/name-of-task/...` in JSON pointers/path. I'd hate to be forced to use indexes again to map objects within the DSL.

wait would that mean using an object for tasks again?

```yaml
do:
  task1: {}
  task2: {}
```

or having a name proerty?

do:
  - name: task1
  - name: task2
```

**Comment by fjtirado** [comment]
I do not think what @ricardozanini is asking it is feasible using an array

**Comment by fjtirado** [comment]
@matthias-pichler-warrify
I think we can avoid additionalproperties in the first schema, but Im not 100% sure  (anyway a minor thing)

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify I think we can avoid additionalproperties in the first schema, but Im not 100% sure (anyway a minor thing)

Don't think so because `properties` can only list named properties. Alternative are `unevaluatedProperties` or `patternProperties` if we fancy 😅

**Comment by fjtirado** [comment]
It seems that once we swith to array, the only way to refer to the object in the array is through index
https://stackoverflow.com/questions/41171075/select-an-array-member-by-name-with-a-json-pointer
Therefore, I think we should go for
```
do: 
   -try: 

   -call:

   -cry: 
```

**Comment by matthias-pichler-warrify** [comment]
> It seems that once we swith to array, the only way to refer to the object in the array is through index https://stackoverflow.com/questions/41171075/select-an-array-member-by-name-with-a-json-pointer Therefore, I think we should go for
> 
> ```
> do: 
>    -try: 
> 
>    -call:
> 
>    -cry: 
> ```
> 
> Since using the name of the task as property itself is just complicating the schema with not percieved value other than having to asign a name to each task (which for people bad for naming like myself is really a pain)

So you propose an optional name property alongside `input`, `output`, `then` ... ? Sounds good to me 👍

**Comment by fjtirado** [comment]
@matthias-pichler-warrify yes, and some day maybe jsonpointer allow filtering using that property if present

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify yes, and some day maybe jsonpointer allow filtering using that property if present

one can dream I guess 🙈 Not to complicate things but I think json-path would allow that ... I prefer json-pointer though 👍 

```
$.do[?(@.name == "task1")]
```

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify I think we should open other topic to discuss the usage of then

Yes that can be moved elsewhere now 👍 My proposal tried to attack two problems: fallthrough and key order. #882 resolves the key order issue (which was the bigger issue anyway)

**Comment by ricardozanini** [comment]
@matthias-pichler-warrify I'm trying to catch up with the discussions, so I can do my homework and explore the technical issue here.

The core problem is that maps order is not guaranteed in JS/Python? And one MUST implement their own parser if they need ordering?

I think we can solve this in the spec documentation instead of adding arrays to the `do`. Having to manipulate indexes in JSON Path/Pointer is too much of a pain and hard to follow. Mainly because we have `errors` that rely on those pointers to be well-defined.

We can even propose and create a CI validating task order on the most popular languages.

**Comment by matthias-pichler-warrify** [comment]
> The core problem is that maps order is not guaranteed in JS/Python? And one MUST implement their own parser if they need ordering?

Yes order is not guaranteed. They are even specifically defined as "unordered". And some languages such as JavaScript even have a defined process to reorder keys. 

Additionally if one wants to store the definition as native json so that one can filter in the database many databases reorder as well: some that come to mind: DynamoDB, MongoDB, jsonb in Postgres, ...

So the problem is not only programming languages but also databases.

We want to provide an API (and do not have an SDK) to create workflows. So now we have to send the definition as a string to rely on order but realistically then also ALL our clients would have to make sure that order is preserved (which `JSON.stringify()` does not do for example).

> I think we can solve this in the spec documentation instead of adding arrays to the do. Having to manipulate indexes in JSON Path/Pointer is too much of a pain and hard to follow. Mainly because we have errors that rely on those pointers to be well-defined.

If json pointers to tasks are of importance then I propose to implement a flow that does not rely on task order. By a required `start` key and required `then` in every task. Then order doesn't matter anymore.

**Comment by ricardozanini** [comment]
Understood. So we need to:

1. Make JSON pointers/paths friendly and easy to write and understand. The main pain point here is the `errors` definition where the implementation MUST adhere to an actual task pointer in the DSL
2. Workflow definitions must be stored, serialized, deserialized, and processed in a way that the desired original order IS PRESERVED.

Assuming my premises are true, I understand that we have to compromise on either having pointers to indexes or having `then` and `start` as required fields.

Does anyone else see another solution? If not, I propose a vote:

- [ ] Use arrays, so order are guaranteed
- [ ] Use maps, but add `then` and `start` required properties to the `task` and `workflow` respectively.

**Comment by matthias-pichler-warrify** [comment]
> Understood. So we need to:
> 
> 1. Make JSON pointers/paths friendly and easy to write and understand. The main pain point here is the `errors` definition where the implementation MUST adhere to an actual task pointer in the DSL
> 2. Workflow definitions must be stored, serialized, deserialized, and processed in a way that the desired original order IS PRESERVED.
> 
> Assuming my premises are true, I understand that we have to compromise on either having pointers to indexes or having `then` and `start` as required fields.
> 
> Does anyone else see another solution? If not, I propose a vote:
> 
> * [ ]  Use arrays, so order are guaranteed
> * [x]  Use maps, but add `then` and `start` required properties to the `task` and `workflow` respectively.

I don't see an alternative.

my vote would be:
> Use maps, but add then and start required properties to the task and workflow respectively.

**Comment by matthias-pichler-warrify** [comment]
@ricardozanini @fjtirado @cdavernas 

The two examples in: https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2145719884 should illustrate the two approaches. I would prefer the array

**Comment by ricardozanini** [comment]
Why not map in tasks and arrays in every other place that makes sense having an array? Array with a property `name` rewinds back to the old style which is one of the sole purposes to present a new DSL. 

Task **must** have a name, but declaring it directly in the task property seems clunky. 

The other examples: `try`, `do` in `for` and `catch`, `before`, and `after` must be reevaluated. I can see a few comments in the PR, let's try to have the discussion here and use the PR once we have a consensus. I think it will organize the discussion.

For those, we can use an array without the `name` property.

**Comment by matthias-pichler-warrify** [comment]
> Why not map in tasks and arrays in every other place that makes sense having an array? Array with a property `name` rewinds back to the old style which is one of the sole purposes to present a new DSL.
> 
> Task **must** have a name, but declaring it directly in the task property seems clunky.
> 
> The other examples: `try`, `do` in `for` and `catch`, `before`, and `after` must be reevaluated. I can see a few comments in the PR, let's try to have the discussion here and use the PR once we have a consensus. I think it will organize the discussion.
> 
> For those, we can use an array without the `name` property.

Yeah the discussion in the PR was about whether `try`, `do` in `for` and `catch`, `before`, and `after` should even accept multiple tasks because what purpose would the `composite.sequential` task serve at this point.

I don't really care about whether these (`try`, `do` in `for` and `catch`, `before`, and `after`) are arrays or not. 

I think however it should be possible to give every task a name!

Either via a name property:

```yaml
- name: getPet
  call: http
  with: {}
```

or a key

```yaml
- getPet:
    call: http
    with: {}
```

I do not have a strong opinion on either of these, I slightly prefer the second (name as key) though.

**Comment by cdavernas** [comment]
> The other examples: try, do in for and catch, before, and after must be reevaluated. I can see a few comments in the PR, let's try to have the discussion here and use the PR once we have a consensus. I think it will organize the discussion.

@ricardozanini I disagree: this PR is about fixing a bug, meaning changing maps into arrays to ensure ordering on every and all platforms. The consensus should be reached on a dedicated issue, that addresses that opiniated change. Said PR already exists btw: #869. I'm however strongly against it, as it makes these statements look clunky at best, and defeats the `composite` task.

> I do not have a strong opinion on either of these, I slightly prefer the second (name as key) though.

Me too!

**Comment by cdavernas** [comment]
Alternative to:

```yaml
  handleUnknownOrderType:
    for:
      each: order
      in: .orders
      start: logWarning
      do:
        logWarning:
          set:
            log: warn
          then: notifyAdmin
        notifyAdmin:
          try:
            start: setMessage
            do:
              setMessage:
                set:
                  message: something's wrong
                then: logMessage
              logMessage:
                set:
                  message2: ${ .message }
                then: exit
          then: exit
```

is what exists for now:

```yaml
  handleUnknownOrderType:
    for:
      each: order
      in: .orders
      start: logWarning
      do:
        logWarning:
          set:
            log: warn
          then: notifyAdmin
        notifyAdmin:
          try:
            do:
              set:
                message: something's wrong
```

No cumbersome `then` and `start` all over the place. Arrays would be applied to the top-level `do` and to the tasks a composite task is made out of.

**Comment by cdavernas** [comment]
@matthias-pichler-warrify Yeah, I was fixing when you answered, sorry.

**Comment by matthias-pichler-warrify** [comment]
So it would be the 

> do with array rest as composite

approach 👍 I also like that one

```yaml
document:
  dsl: "1.0.0-alpha1"
  namespace: test
  name: do-as-an-array-composite-elsewhere
  version: "0.1.0"
use:
  extensions:
    - externalLogging:
        extend: all
        before:
          sendLog:
            call: http
            with:
              method: post
              uri: https://fake.log.collector.com
              body:
                message: msg
        after:
          sendLog:
            call: http
            with:
              method: post
              uri: https://fake.log.collector.com
              body:
                message: msg
do:
  - processOrder:
      switch:
        - case1:
            when: .orderType == "electronic"
            then: processElectronicOrder
        - case2:
            when: .orderType == "physical"
            then: processPhysicalOrder
        - default:
            then: handleUnknownOrderType
  - processElectronicOrder:
      execute:
        sequentially:
          - validatePayment:
              set:
                validate: true
          - fulfillOrder:
              set:
                status: fulfilled
        then: exit
  - processPhysicalOrder:
      execute:
        concurrently:
          - checkInventory:
              set:
                inventory: clear
          - packItems:
              set:
                items: 1
          - scheduleShipping:
              set:
                address: Elmer St
        then: exit
  - handleUnknownOrderType:
      for:
        each: order
        in: .orders
        do:
          execute:
            sequentially:
              - logWarning:
                  set:
                    log: warn
              - notifyAdmin:
                  try:
                    - setMessage:
                        set:
                          message: something's wrong
                    - logMessage:
                        set:
                          message2: ${ .message }
```

**Comment by cdavernas** [comment]
@matthias-pichler-warrify Yes, exactly! It's my personnal favorite as it retains a fluent and linguistic approach. WDYT?

It would fix the bug, without changing the actual structure. 

Like said above, if we feel we should change the structure of try, etc., I believe it should be addressed in dedicated issue, and should reach a consensus, which I'm not feeling for now.

**Comment by matthias-pichler-warrify** [comment]
@cdavernas I agree and I think it's what I have in #882 so far but I will check so the PR is ready if we reach consensus on this 👍

**Comment by cdavernas** [comment]
@matthias-pichler-warrify you rule! Thanks!

**Comment by fjtirado** [comment]
Im NOT ok with forcing names in arrays. It complicates the schema unnecesarily and I hardly see any benefit other than the writer having to name the task. 
I cannot also see how putting names make the workflow more fluent ( I would say it makes it more verbose). In my opinion Names should be optional. 
And yes, contrary to the do: in loops  ( which is a question of internal coherence) this is my personal opinion and Im willing to compromise, so lets go with mandatory names (but in exchange, please make the do: schema consistenct at least)

**Comment by matthias-pichler-warrify** [comment]
> Im NOT ok with forcing names in arrays. It complicates the schema unnecesarily and I hardly see any benefit other than the writer having to name the task.

my thought was that forcing names:
- makes it better to visualize (named blocks for tasks)
- easier to reference in `then` because everything is already named
- task names can appear in logs or friendly error messages

> I cannot also see how putting names make the workflow more fluent ( I would say it makes it more verbose). In my opinion Names should be optional.

I am fine either way, although I prefer the name as the key approach. It is more verbose but I do find it very readable personally 👍 

> And yes, contrary to the do: in loops ( which is a question of internal coherence) this is my personal opinion and Im willing to compromise, so lets go with mandatory names (but in exchange, please make the do: schema consistenct at least, so in any place where do: currently appears, lets use an array of task)

As said before, if you are willing to compromise on the mandatory names, I am ok with making `for.do` an array

**Comment by cdavernas** [comment]
Should we poll on that? I'm fine with whatever we decide.

My personnal opinion is that the alternative to using a key is using an optional `name` property, which is not fluent and which will likely never be set anyways, since it's optional. I'd therefore use keys, even though I agree with @fjtirado that it's  purely cosmetic (and will be redundant in references: `/do/0/log` == `/do/0`)

**Comment by matthias-pichler-warrify** [comment]
> Should we poll on that? I'm fine with whatever we decide.
> 
> My personnal opinion is that the alternative to using a key is using an optional `name` property, which is not fluent and which will likely never be set anyways, since it's optional. I'd therefore use keys, even though I agree with @fjtirado that it's purely cosmetic (and will be redundant in references: `/do/0/log` == `/do/0`)
> 
> > my thought was that forcing names:
> > makes it better to visualize (named blocks for tasks)
> > easier to reference in then because everything is already named
> > task names can appear in logs or friendly error messages
> 
> I totally agree with you

In that case I think my vote would be: "names as keys" and `for.do` being an array to reach consensus on this 👍

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
I also vote for names as keys and for..do as an array for compromise as far as they are in the same PR (not deferring the for..do as an array to a different one)

**Comment by matthias-pichler-warrify** [comment]
In the meantime I updated #882 to make `for.do` an array as well. (even added a new `$def/do` to reuse the exact same schema)

**Comment by cdavernas** [comment]
> In the meantime I updated https://github.com/serverlessworkflow/specification/pull/882 to make for.do an array as well

No, I'm not. Because then we have the useless `composite.sequentially` that needs to be addressed. That's why it needs to be addressed in another issue, if and when consensus is reached.

And I vote for keys too.

**Comment by matthias-pichler-warrify** [comment]
> > In the meantime I updated #882 to make for.do an array as well
> 
> No, I'm not. Because then we have the useless `composite.sequentially` that needs to be addressed. That's why it needs to be addressed in another issue, if and when consensus is reached.
> 
> Also no because the issue at hand can be resolved by simply renaming for/do to something else.
> 
> And I vote for keys too.

ok then I'll revert that change 👍 

`composite.sequentially` would still be useful for `try`, `catch`, `before` & `after`

**Comment by cdavernas** [comment]
> composite.sequentially would still be useful for try, catch, before & after

Sorry, I missed that because of #869.

I'd rather rename `for/do`, for a purely cosmetic concern.
I'm also afraid we will have to add properties to a sequential action list, in the future, and we will therefore have to "pollute" the for to do so.

**Comment by fjtirado** [comment]
@cdavernas  Another possibility is to completely remove do: from everywhere and just use composition in any place where you need arrays of task (including main flow)

**Comment by fjtirado** [comment]
Note that do: is now used at main flow, in loops and catch

**Comment by cdavernas** [comment]
Good point, but I think you'll agree that it would completly defeat the fluent and linguistic approach to the workflow; read out loud it won't mean anything anymore.

Whereas as just renaming /for/do seems the less painful.

**Comment by matthias-pichler-warrify** [comment]
> @cdavernas Another possibility is to completely remove do: from everywhere and just use composition in any place where you need arrays of task (including main flow)

you mean to basically replace `do` with `execute`:

```yaml
document:
  dsl: "1.0.0-alpha1"
  namespace: test
  name: execute-all-the-things
  version: "0.1.0"
use:
  extensions:
    - externalLogging:
        extend: all
        before:
          call: http
          with:
            method: post
            uri: https://fake.log.collector.com
            body:
              message: msg
        after:
          call: http
          with:
            method: post
            uri: https://fake.log.collector.com
            body:
              message: msg
execute:
  sequentially:
    - processOrder:
        switch:
          - case1:
              when: .orderType == "electronic"
              then: processElectronicOrder
          - case2:
              when: .orderType == "physical"
              then: processPhysicalOrder
          - default:
              then: handleUnknownOrderType
    - processElectronicOrder:
        execute:
          sequentially:
            - validatePayment:
                set:
                  validate: true
            - fulfillOrder:
                set:
                  status: fulfilled
          then: exit
    - processPhysicalOrder:
        execute:
          concurrently:
            - checkInventory:
                set:
                  inventory: clear
            - packItems:
                set:
                  items: 1
            - scheduleShipping:
                set:
                  address: Elmer St
          then: exit
    - handleUnknownOrderType:
        for:
          each: order
          in: .orders
          execute:
            sequentially:
              - logWarning:
                  set:
                    log: warn
              - notifyAdmin:
                  try:
                    execute:
                      sequentially:
                        - setMessage:
                            set:
                              message: something's wrong
                        - logMessage:
                            set:
                              message2: ${ .message }
```

**Comment by cdavernas** [comment]
> Note that do: is now used at main flow, in loops and catch

Right! So why not leaving `do` at top level to keep fluency, and remove it from, while and catch?

**Comment by cdavernas** [comment]
> you mean to basically replace do with execute:

Yeah, why not! That's also a pretty good idea. Even though it might be a tad more cumbersome when wanting only a single task.

**Comment by matthias-pichler-warrify** [comment]
> > you mean to basically replace do with execute:
> 
> Yeah, why not! That's also a pretty good idea. Even though it might be a tad more cumbersome when wanting only a single task.

true but do I really need a work _flow_ for a single task? 😅

**Comment by cdavernas** [comment]
> true but do I really need a work flow for a single task? 😅

Workflow no. However try/for/... fosho!

**Comment by matthias-pichler-warrify** [comment]
> > true but do I really need a work flow for a single task? 😅
> 
> Workflow no. However try/for/... fosho!

touché ... what if we introduced a shorthand to write `execute.sequentially: map[string, task][]` with `execute: map[string, task][]`

**Comment by fjtirado** [comment]
@matthias-pichler-warrify  @JBBianchi Just do that here https://github.com/serverlessworkflow/specification/discussions/884#discussioncomment-9662801
I lile @JBBianchi proposal, which I think adress the logical root cause of the issue, so +1 for that.

**Comment by cdavernas** [comment]
> touché ... what if we introduced a shorthand to write execute.sequentially: map[string, task][] with execute: map[string, task][]

Because of OneOfs. It's something that is a huge PITA for us OO devs (isn't it @fjtirado ?). I'd rather go with @JBBianchi's proposal in #884

**Comment by fjtirado** [comment]
so basically we do not have execute, we have do: for single task and concurrently or sequentially for multi task

**Comment by matthias-pichler-warrify** [comment]
> so basically we do not have execute, we have do: for single task and concurrently or sequentially for multi task. That I think works for everyone. Lets see it reflected on the schema ;)

perfect I'll get to it 👍

**Comment by JBBianchi** [comment]
I think we overlooked the case of `switch` with the adopted solution. It's relegated to a second zone citizen, only available in `execute`:
```yaml
do:
  execute:
    sequentially:
      - processOrder:
          switch:
          # ...
```
https://github.com/serverlessworkflow/specification/blob/main/dsl-reference.md#switch

Shouldn't we go back to `do` being a `map[string, task][]` instead of a single task ?

**Comment by cdavernas** [comment]
@JBBianchi Yes, indeed. Do as a mapping array is what we should opt for:

- It resolves consistency of the keyword `do`: it's alway a mapping array of tasks
- It allows for top-level switch statements
- It's more consistent with what a workflow is, to most people: a successions of multiple tasks, not a single top level one, which is then... a task

**Comment by matthias-pichler-warrify** [comment]
That is true. And `do` as an array is absolutely fine by me, however as @cdavernas pointed out: what's the purpose of `sequentially` then? 

Which given the discussion around `concurrently` and the `execute` keyword in general, might mean:

- drop sequentially (not needed as we have arrays everywhere 
- only keep `concurrently` (or replace with `branch` to make it a verb for consistency)

**Comment by cdavernas** [comment]
> only keep concurrently (or replace with branch to make it a verb for consistency)

That's a good idea. Even though I like `branch`, I'm not sure it conveys the idea of running concurrently :(. `multitask`? Alternatives?

**Comment by matthias-pichler-warrify** [comment]
> That's a good idea. However, even though I like `branch`, I'm not sure it conveys the idea of running concurrently 😞. 
> 
> `multitask`, or something like that?

Hmm `multitask` certainly isn't consistent or a verb 😅 I'd prefer  keeping `concurrently` over it.

The idea for `branch` also came from AWS StepFunctions (they run in parallel there)

**Comment by cdavernas** [comment]
> multitask
/ˈmʌltɪtɑːsk/
verb 
> 1. (of a person) deal with more than one task at the same time.
"I managed my time efficiently and multitasked"

It is a verb, though certainly not the best term. But there is only a little amount of alternatives, such as the (horrible) `parallelize`

**Comment by JBBianchi** [comment]
Do you mean:
```yaml
do:
  execute:
    branch:
      - task1:
      - task2:
      - task3:    
```

or
```yaml
do:
  branch:
    - task1:
    - task2:
    - task3: 
```

(an alternative to `branch` could also be `fork`)

In the first case, we keep the "controversial" `execute` but have an extra level to put config like `compete: true`. In the second one, we get rid of the controversial keyword but we might need an extra level of depth.

**Comment by fjtirado** [comment]
@JBBianchi 
Im trying to undestand the issue with swith. I understood that rather than writing
```
do:
  execute:
    sequentially:
      - processOrder:
          switch:
            - case1:
                when: .orderType == "electronic"
                then: processElectronicOrder
            - case2:
                when: .orderType == "physical"
                then: processPhysicalOrder
            - default:
                then: handleUnknownOrderType
      - processElectronicOrder:
          execute:
            sequentially:
              - validatePayment: {}
              - fulfillOrder: {}
          then: exit
      - processPhysicalOrder:
          execute:
            sequentially:
              - checkInventory: {}
              - packItems: {}
              - scheduleShipping: {}
          then: exit
      - handleUnknownOrderType:
          execute:
            sequentially:
              - logWarning: {}
              - notifyAdmin: {}
```

we should be able to write
```
do:
     - processOrder:
          switch:
            - case1:
                when: .orderType == "electronic"
                then: processElectronicOrder
            - case2:
                when: .orderType == "physical"
                then: processPhysicalOrder
            - default:
                then: handleUnknownOrderType
      - processElectronicOrder:
          execute:
            sequentially:
              - validatePayment: {}
              - fulfillOrder: {}
          then: exit
      - processPhysicalOrder:
          execute:
            sequentially:
              - checkInventory: {}
              - packItems: {}
              - scheduleShipping: {}
          then: exit
      - handleUnknownOrderType:
          execute:
            sequentially:
              - logWarning: {}
              - notifyAdmin: {}
```

isnt it?
But it we remove execute, as option 2 in this [discussion](https://github.com/serverlessworkflow/specification/discussions/889), it would be   

```
do:
  sequentially:
     - processOrder:
          switch:
            - case1:
                when: .orderType == "electronic"
                then: processElectronicOrder
            - case2:
                when: .orderType == "physical"
                then: processPhysicalOrder
            - default:
                then: handleUnknownOrderType
      - processElectronicOrder:
          execute:
            sequentially:
              - validatePayment: {}
              - fulfillOrder: {}
          then: exit
      - processPhysicalOrder:
          execute:
            sequentially:
              - checkInventory: {}
              - packItems: {}
              - scheduleShipping: {}
          then: exit
      - handleUnknownOrderType:
          execute:
            sequentially:
              - logWarning: {}
              - notifyAdmin: {}
```
Am I missing anything? (probably I am, my apologies in advance)

**Comment by JBBianchi** [comment]
> @JBBianchi Im trying to undestand the issue with swith. I understood that rather than writing
> 
> ```
> do:
>   execute:
>     sequentially:
>       - processOrder:
>           switch:
>             - case1:
>                 when: .orderType == "electronic"
>                 then: processElectronicOrder
>             - case2:
>                 when: .orderType == "physical"
>                 then: processPhysicalOrder
>             - default:
>                 then: handleUnknownOrderType
>       - processElectronicOrder:
>           execute:
>             sequentially:
>               - validatePayment: {}
>               - fulfillOrder: {}
>           then: exit
>       - processPhysicalOrder:
>           execute:
>             sequentially:
>               - checkInventory: {}
>               - packItems: {}
>               - scheduleShipping: {}
>           then: exit
>       - handleUnknownOrderType:
>           execute:
>             sequentially:
>               - logWarning: {}
>               - notifyAdmin: {}
> ```
> 
> we should be able to write
> 
> ```
> do:
>      - processOrder:
>           switch:
>             - case1:
>                 when: .orderType == "electronic"
>                 then: processElectronicOrder
>             - case2:
>                 when: .orderType == "physical"
>                 then: processPhysicalOrder
>             - default:
>                 then: handleUnknownOrderType
>       - processElectronicOrder:
>           execute:
>             sequentially:
>               - validatePayment: {}
>               - fulfillOrder: {}
>           then: exit
>       - processPhysicalOrder:
>           execute:
>             sequentially:
>               - checkInventory: {}
>               - packItems: {}
>               - scheduleShipping: {}
>           then: exit
>       - handleUnknownOrderType:
>           execute:
>             sequentially:
>               - logWarning: {}
>               - notifyAdmin: {}
> ```
> 
> isnt it? 

That's my concern indeed.

> But it we remove execute, as option 2 in this [discussion](https://github.com/serverlessworkflow/specification/discussions/889), it would be
> 
> ```
> do:
>   sequentially:
>      - processOrder:
>           switch:
>             - case1:
>                 when: .orderType == "electronic"
>                 then: processElectronicOrder
>             - case2:
>                 when: .orderType == "physical"
>                 then: processPhysicalOrder
>             - default:
>                 then: handleUnknownOrderType
>       - processElectronicOrder:
>           execute:
>             sequentially:
>               - validatePayment: {}
>               - fulfillOrder: {}
>           then: exit
>       - processPhysicalOrder:
>           execute:
>             sequentially:
>               - checkInventory: {}
>               - packItems: {}
>               - scheduleShipping: {}
>           then: exit
>       - handleUnknownOrderType:
>           execute:
>             sequentially:
>               - logWarning: {}
>               - notifyAdmin: {}
> ```
> 
> Am I missing anything? (probably I am, my apologies in advance)

The later example doesn't seem to match any late suggestions made in this issue. I didn't mention removing `execute` per se as it's another topic. But if I understand what @matthias-pichler-warrify and @cdavernas were talking about just after I raised my concern, is that `sequentially` is the default behavior of a `do`, therefore `execute.sequentially` doesn't really have a reason to exists. Which leaves us with `concurrently`, which could be replaced by `branch`(, `multitask`, `parallelize` or `fork`). What's unclear is if we keep `execute.branch` (or alternative) or we just get rid of `execute` altogether an opt for `branch` (or alternative) straight on - which then, rejoin the point 2 of #889 and makes the discussion obsolete. In the later case though, we'd probably need an extra level of depth for some params like racing (or threads count for instance).

e.g.:
```yaml
do:
  - processOrder:
      switch:
        - case1:
            when: .orderType == "electronic"
            then: processElectronicOrder
        - case2:
            when: .orderType == "physical"
            then: processPhysicalOrder
        - default:
            then: handleUnknownOrderType
  - processElectronicOrder:
      do:
        - validatePayment: {}
        - fulfillOrder: {}
        then: exit
  - processPhysicalOrder:
      execute:
        compete: true
        branch:  # `multitask`, `parallelize` or `fork`
          - checkInventory: {}
          - packItems: {}
          - scheduleShipping: {}
            then: exit
  - handleUnknownOrderType: #...
```

or
```yaml
do:
  - processOrder:
      switch:
        - case1:
            when: .orderType == "electronic"
            then: processElectronicOrder
        - case2:
            when: .orderType == "physical"
            then: processPhysicalOrder
        - default:
            then: handleUnknownOrderType
  - processElectronicOrder:
      do:
        - validatePayment: {}
        - fulfillOrder: {}
        then: exit
  - processPhysicalOrder:
      branch:
        compete: true
        on:  # `multitask`, `parallelize` or `fork`
          - checkInventory: {}
          - packItems: {}
          - scheduleShipping: {}
            then: exit
  - handleUnknownOrderType: #...
```

@cdavernas @matthias-pichler-warrify am I getting it right ?

**Comment by ricardozanini** [comment]
Guys, what about just focus on solving the problem that was oversight and making the tasks an array? Then we open another issue to tackle the execute/concurrently debate.

It's already hard to follow up on the discussion on this issue. What about a PR to solve the `do:` in a single task, we close this one and open another?

**Comment by fjtirado** [comment]
@JBBianchi There was a typo in my last example, I edited it. My point was that this is kind of related with the solution for do: execute:, if we simplify how multitask is written, its not really an issue that switch is a task which only makes sense within a multitask (which is why we are proposing to make do: multitask again)

**Comment by cdavernas** [comment]
> its not really an issue that switch is a task which only makes sense within a multitask (which is why we are proposing to make do: multitask again)

I strongly disagree: it is a problem, and a big one. 
The fact that `do` could have different meaning is blocking for you in terms of consistency, but not the fact that the `switch` task  - and the `switch` task only - cannot be used at top level?
Furthermore, as said above, a single task per workflow goes against the definition of a workflow. It also is a problem when it comes to visualization.

> (which is why we are proposing to make do: multitask again)

That is not what we are proposing. Also, in this issue, we are ONLY proposing to make do an array everywhere again. All other issues/concerns should be moved to dedicated issue or PR, as @ricardozanini said.

**Comment by fjtirado** [comment]
@cdavernas 
I think we are are saying the same thing, so there is not need to strongly disagree ;)
When I said that we are proposing to do a multitask I mean do: being an array (which I said is fine as far as do: have the same syntax everywhere, not only in main flow) 
And removing execute/sequentially is part of the do:execute: redundancy discussion.

**Comment by cdavernas** [comment]
> I think we are are saying the same thing, so there is not need to strongly disagree ;)

Hehe, sorry, seems I misunderstood you indeed 🤣 

> And removing execute/sequentially is part of the do:execute: redundancy discussion.

Indeed.
